pytorch_lightning/profiler/pytorch.py

```diff
@@ -0,0 +1,681 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Profiler to check if there are any bottlenecks in your code."""
+import inspect
+import os
+from functools import partial
+from typing import Any, Callable, List, Optional
+
+import torch
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.profiler.profilers import BaseProfiler
+from pytorch_lightning.utilities import _TORCH_GREATER_EQUAL_1_8, rank_zero_only
+from pytorch_lightning.utilities.cloud_io import get_filesystem
+from pytorch_lightning.utilities.distributed import rank_zero_warn
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _TORCH_GREATER_EQUAL_1_8:
+    from torch.autograd.profiler import record_function
+    from torch.profiler import ProfilerAction, ProfilerActivity, tensorboard_trace_handler
+
+
+class RegisterRecordFunction:
+    """
+    While profiling autograd operations, this class will add label with module name
+    around the forward function.
+
+    The Lightning PyTorch Profiler will activate this feature automatically.
+
+    It can be desactivated as follow:
+
+    Example::
+
+        from pytorch_lightning.profilers import PyTorchProfiler
+
+        profiler = PyTorchProfiler(record_module_names=False)
+
+        Trainer(profiler=profiler)
+
+    It can be used outside of Lightning as follow:
+
+    Example::
+
+        from pytorch_lightning import Trainer, seed_everything
+
+        with RegisterRecordFunction(model):
+            out = model(batch)
+
+    """
+
+    def __init__(self, model):
+        self._model = model
+        self._records = {}
+        self.handles = {}
+
+    def _start_recording(self, module, input, module_name: str = None, is_built_in: bool = None):
+        if module_name is not None:
+            record_name = module_name if is_built_in else f"{type(module)}: {module_name}"
+            self._records[record_name] = record_function(record_name).__enter__()
+        return input
+
+    def _stop_recording(self, module, input, result, module_name: str = None, is_built_in: bool = None):
+        if module_name is not None:
+            record_name = module_name if is_built_in else f"{type(module)}: {module_name}"
+            self._records[record_name].__exit__(None, None, None)
+        return result
+
+    def __enter__(self):
+        built_in_modules = dir(torch.nn)
+        for module_name, module in self._model.named_modules():
+            is_built_in = module in built_in_modules
```

Comment:
is this expensive for each `__enter__`?

Reply:
This enter will be done only once during the entire recording.